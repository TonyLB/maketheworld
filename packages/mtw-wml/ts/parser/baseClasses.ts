import {
    Token,
    TokenExpressionValue,
    TokenKeyValue,
    TokenLiteralValue,
    ParseStackTokenEntry
} from './tokenizer/baseClasses'

type ParseTagBase = {
    startTagToken: number;
    endTagToken: number;
}

export type ParseLiteralLegalContents = ParseStringTag
type ParseValueBase = {
    value: string;
    contents: ParseLiteralLegalContents[];
} & ParseTagBase

export type ParseAssetLegalContents = ParseActionTag |
    ParseCommentTag |
    ParseComputedTag |
    ParseConditionTag |
    ParseExitTag |
    ParseFeatureTag |
    ParseImageTag |
    ParseImportTag |
    ParseMapTag |
    ParseRoomTag |
    ParseVariableTag |
    ParseWhitespaceTag

type ParseAssetBase = {
    key: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    player?: string;
    contents: ParseAssetLegalContents[];
} & ParseTagBase

export type ParseAssetTag = {
    tag: 'Asset';
} & ParseAssetBase

export type ParseStoryTag = {
    tag: 'Story';
    instance: boolean;
} & ParseAssetBase

export type ParseNameTag = ParseTaggedMessageTag<"Name">

export type ParsePronounsTag = {
    tag: 'Pronouns';
    subject: string;
    object: string;
    possessive: string;
    reflexive: string;
    adjective: string;
} & ParseTagBase

export type ParseOneCoolThingTag = {
    tag: 'OneCoolThing';
} & ParseValueBase

export type ParseFirstImpressionTag = {
    tag: 'FirstImpression';
} & ParseValueBase

export type ParseOutfitTag = {
    tag: 'Outfit';
} & ParseValueBase

export type ParseImageTag = {
    tag: 'Image';
    key: string;
    fileURL?: string;
} & ParseTagBase

export type ParseCharacterLegalContents = ParseNameTag | ParsePronounsTag | ParseFirstImpressionTag | ParseOneCoolThingTag | ParseOutfitTag | ParseImageTag
export type ParseCharacterTag = {
    tag: 'Character';
    key: string;
    player?: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    contents: ParseCharacterLegalContents[];
} & ParseTagBase

export type ParseVariableTag = {
    tag: 'Variable';
    key: string;
    default?: string;
} & ParseTagBase

export type ParseDependencyTag = {
    tag: 'Depend';
    on: string;
} & ParseTagBase

export type ParseComputedTag = {
    tag: 'Computed';
    key: string;
    src: string;
    dependencies: ParseDependencyTag[];
} & ParseTagBase

export type ParseActionTag = {
    tag: 'Action';
    key: string;
    src: string;
} & ParseTagBase

export type ParseUseTag = {
    tag: 'Use';
    key: string;
    as?: string;
    type?: string;
} & ParseTagBase

export type ParseImportLegalContents = ParseUseTag
export type ParseImportTag = {
    tag: 'Import';
    from: string;
    contents: ParseImportLegalContents[];
} & ParseTagBase

export type ParseConditionLegalContextTag = 'Asset' | 'Description' | 'Room' | 'Feature' | 'Map'

export type ParseConditionTagAssetContext = {
    tag: 'If';
    contextTag: 'Asset';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseAssetLegalContents[];
} & ParseTagBase

export type ParseConditionTagDescriptionContext = {
    tag: 'If';
    contextTag: 'Description';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseTaggedMessageLegalContents[];
} & ParseTagBase

export type ParseConditionTagRoomContext = {
    tag: 'If';
    contextTag: 'Room';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseRoomLegalContents[];
} & ParseTagBase

export type ParseConditionTagFeatureContext = {
    tag: 'If';
    contextTag: 'Feature';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseFeatureLegalContents[];
} & ParseTagBase

export type ParseConditionTagMapContext = {
    tag: 'If';
    contextTag: 'Map';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseMapLegalContents[];
} & ParseTagBase

export type ParseConditionTypeFromContextTag<T extends ParseConditionLegalContextTag> =
    T extends 'Description'
        ? ParseConditionTagDescriptionContext
        : T extends 'Room'
            ? ParseConditionTagRoomContext
            : T extends 'Feature'
                ? ParseConditionTagFeatureContext
                : T extends 'Map'
                    ? ParseConditionTagMapContext
                    : ParseConditionTagAssetContext

export type ParseConditionTag = ParseConditionTagAssetContext | ParseConditionTagDescriptionContext | ParseConditionTagRoomContext | ParseConditionTagFeatureContext | ParseConditionTagMapContext

export const isLegalParseConditionContextTag = (value: string): value is ParseConditionLegalContextTag => (['Asset', 'Description', 'Room', 'Feature', 'Map'].includes(value))
export const isParseConditionTagAssetContext = (value: ParseConditionTag): value is ParseConditionTagAssetContext => (value.contextTag === 'Asset')
export const isParseConditionTagDescriptionContext = (value: ParseConditionTag): value is ParseConditionTagDescriptionContext => (value.contextTag === 'Description')
export const isParseConditionTagRoomContext = (value: ParseConditionTag): value is ParseConditionTagRoomContext => (value.contextTag === 'Room')
export const isParseConditionTagFeatureContext = (value: ParseConditionTag): value is ParseConditionTagFeatureContext => (value.contextTag === 'Feature')
export const isParseConditionTagMapContext = (value: ParseConditionTag): value is ParseConditionTagMapContext => (value.contextTag === 'Map')

export const parseDifferentiatingTags: Record<ParseConditionLegalContextTag,  ParseTag["tag"][]> = {
    Asset: ['Exit', 'Feature', 'Room', 'If', 'Image', 'Map'],
    Description: ['If', 'Space', 'String', 'Link', 'br', 'Whitespace'],
    Room: ['If', 'Description', 'Name', 'Exit'],
    Feature: ['If', 'Description', 'Name'],
    Map: ['If', 'Image', 'Room', 'Name', 'Exit']
}

export type ParseExitTag = {
    tag: 'Exit';
    key?: string;
    name: string;
    to?: string;
    from?: string;
    contents: ParseStringTag[];
} & ParseTagBase

export type ParseLinkLegalContents = ParseStringTag | ParseWhitespaceTag
export type ParseLinkTag = {
    tag: 'Link';
    to: string;
    contents: ParseLinkLegalContents[];
} & ParseTagBase

export type ParseTaggedMessageLegalContents = ParseWhitespaceTag | ParseStringTag | ParseLinkTag | ParseLineBreakTag | ParseSpacerTag | ParseConditionTagDescriptionContext

export type ParseTaggedMessageTag<T extends string> = {
    tag: T;
    display?: 'before' | 'after' | 'replace';
    contents: ParseTaggedMessageLegalContents[];
} & ParseTagBase

export type ParseDescriptionTag = ParseTaggedMessageTag<"Description">

export type ParseLineBreakTag = {
    tag: 'br';
} & ParseTagBase

export type ParseSpacerTag = {
    tag: 'Space';
} & ParseTagBase

export type ParseRoomLegalContents = ParseDescriptionTag | ParseNameTag | ParseExitTag | ParseFeatureTag | ParseConditionTagRoomContext
export type ParseRoomTag = {
    tag: 'Room';
    key: string;
    global: boolean;
    display?: string;
    x?: number;
    y?: number;
    contents: ParseRoomLegalContents[];
} & ParseTagBase

export type ParseFeatureLegalContents = ParseDescriptionTag | ParseNameTag | ParseConditionTagFeatureContext
export type ParseFeatureTag = {
    tag: 'Feature';
    key: string;
    global: boolean;
    contents: ParseFeatureLegalContents[];
} & ParseTagBase

export type ParseMapLegalContents = ParseNameTag | ParseRoomTag | ParseImageTag | ParseExitTag | ParseConditionTagMapContext
export type ParseMapTag = {
    tag: 'Map';
    key: string;
    contents: ParseMapLegalContents[];
} & ParseTagBase

export type ParseStringTag = {
    tag: 'String';
    value: string;
} & ParseTagBase

export type ParseWhitespaceTag = {
    tag: 'Whitespace';
} & ParseTagBase

export type ParseCommentTag = {
    tag: 'Comment';
} & ParseTagBase

export type ParseTag = ParseAssetTag |
    ParseStoryTag |
    ParseNameTag |
    ParseCharacterTag |
    ParsePronounsTag |
    ParseOneCoolThingTag |
    ParseFirstImpressionTag |
    ParseOutfitTag |
    ParseImageTag |
    ParseVariableTag |
    ParseComputedTag |
    ParseDependencyTag |
    ParseActionTag |
    ParseUseTag |
    ParseImportTag |
    ParseConditionTag |
    ParseExitTag |
    ParseDescriptionTag |
    ParseLineBreakTag |
    ParseLinkTag |
    ParseRoomTag |
    ParseFeatureTag |
    ParseMapTag |
    ParseStringTag |
    ParseWhitespaceTag |
    ParseSpacerTag |
    ParseCommentTag

export type ParseLegalTag = ParseTag["tag"]

export const isParseLegalTag = (tag: string): tag is ParseLegalTag => ([
    'Asset',
    'Story',
    'Name',
    'Character',
    'Pronouns',
    'OneCoolThing',
    'FirstImpression',
    'Outfit',
    'Image',
    'Variable',
    'Computed',
    'Depend',
    'Action',
    'Use',
    'Import',
    'If',
    'Exit',
    'Description',
    'br',
    'Link',
    'Room',
    'Feature',
    'Map',
    'String',
    'Whitespace',
    'Space',
    'Comment'
].includes(tag))

export type ParseStackTagOpenPendingEntry = {
    type: 'TagOpenPending';
    tag: string;
    startTagToken: number;
}

export type ParseStackTagOpenEntry = {
    type: 'TagOpen';
    tag: ParseLegalTag;
    startTagToken: number;
    properties: Record<string, (TokenExpressionValue | TokenKeyValue | TokenLiteralValue | boolean)>;
}

export type ParseStackTagEntry<T extends ParseTag> = {
    type: 'Tag';
    tag: T;
}

export class ParseException extends Error {
    startToken: number
    endToken: number
    constructor(message: string, startToken: number, endToken: number) {
        super(message)
        this.name = 'ParseException'
        this.startToken = startToken
        this.endToken = endToken
    }
}

export const isParseTagDependency = (value: ParseTag): value is ParseDependencyTag => (value.tag === 'Depend')
export const isParseTagNesting = (value: ParseTag): value is (ParseRoomTag | ParseFeatureTag | ParseAssetTag | ParseStoryTag | ParseCharacterTag | ParseImportTag | ParseDescriptionTag | ParseConditionTag | ParseLinkTag | ParseMapTag | ParseExitTag | ParseNameTag | ParseFirstImpressionTag | ParseOneCoolThingTag | ParseOutfitTag) => (
    ['Room', 'Feature', 'Asset', 'Story', 'Character', 'Import', 'Description', 'If', 'Link', 'Map', 'Exit', 'Name', 'FirstImpression', 'OneCoolThing', 'Outfit'].includes(value.tag)
)
export const isParseExit = (value: ParseTag): value is ParseExitTag => (value.tag === 'Exit')
export const isParseRoom = (value: ParseTag): value is ParseRoomTag => (value.tag === 'Room')
export const isParseString = (value: ParseTag): value is ParseStringTag => (value.tag === 'String')

export type ParseStackEntry = ParseStackTagOpenPendingEntry | ParseStackTagOpenEntry | ParseStackTagEntry<ParseTag> | ParseStackTokenEntry<Token>

export const isParseStackTagOpenEntry = (value: ParseStackEntry): value is ParseStackTagOpenEntry => (value.type === 'TagOpen')

export type ParseTagFactory<T extends ParseTag> = (value: { open: ParseStackTagOpenEntry, context: ParseStackTagOpenEntry[]; contents: ParseTag[], endTagToken: number }) => ParseStackTagEntry<T>

export const parseTagDefaultProps: Record<string, string> = {
    If: 'if'
}
