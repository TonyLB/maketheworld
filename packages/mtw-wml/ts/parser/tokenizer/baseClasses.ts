import SourceStream from "./sourceStream";

export type TokenBase = {
    source: string;
    startIdx: number;
    endIdx: number;
}

export type TokenComment = {
    type: 'Comment'
} & TokenBase

export type TokenWhitespace = {
    type: 'Whitespace'
} & TokenBase

export type TokenExpressionValue = {
    type: 'ExpressionValue'
} & TokenBase

export type TokenKeyValue = {
    type: 'KeyValue';
    key: string;
} & TokenBase

export type TokenLiteralValue = {
    type: 'LiteralValue';
    value: string;
} & TokenBase

export type TokenProperty = {
    type: 'Property'
} & TokenBase

export type TokenBeginTagOpen = {
    type: 'BeginTagOpen'
} & TokenBase

export type TokenBeginTagClose = {
    type: 'BeginTagClose';
} & TokenBase

export type TokenEndTag = {
    type: 'EndTag';
} & TokenBase

export type TokenDescription = {
    type: 'Description';
} & TokenBase

export type TokenError = {
    type: 'Error';
    message: string;
} & TokenBase

export type Token = TokenComment |
    TokenWhitespace |
    TokenExpressionValue |
    TokenKeyValue |
    TokenLiteralValue |
    TokenProperty |
    TokenBeginTagOpen |
    TokenBeginTagClose |
    TokenEndTag |
    TokenDescription |
    TokenError

export type Tokenizer<T extends TokenBase> = {
    (sourceStream: SourceStream): T | TokenError | undefined
}

export const isTokenError = (item: TokenBase & { type: string }): item is TokenError => (item.type === 'Error')
