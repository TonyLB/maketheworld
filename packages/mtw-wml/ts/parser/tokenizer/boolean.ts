import {
    Tokenizer,
    TokenBooleanProperty
} from "./baseClasses"

export const booleanPropertyTokenizer: Tokenizer<TokenBooleanProperty> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar.match(/[A-Za-z\!]/)) {
        const startIdx = sourceStream.position
        let returnValue = firstChar === '!' ? sourceStream.consume(1) : ''
        let key = ''
        while(sourceStream.lookAhead(1).match(/[A-Za-z]/)) {
            key = key + sourceStream.consume(1)
        }
        returnValue = returnValue + key
        return {
            type: 'BooleanValue',
            source: returnValue,
            value: firstChar !== '!',
            key,
            startIdx,
            endIdx: sourceStream.position - 1
        }
    }
    else {
        return undefined
    }
}

export default booleanPropertyTokenizer
