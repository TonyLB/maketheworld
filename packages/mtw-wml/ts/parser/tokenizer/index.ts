import {
    isTokenError,
    Token,
    TokenDescription,
    TokenTagClose,
    TokenTagOpenBegin,
    TokenTagOpenEnd,
    TokenWhitespace,
    TokenProperty,
    TokenBooleanProperty,
    TokenLiteralValue,
    TokenKeyValue,
    TokenExpressionValue
} from "./baseClasses"
import booleanPropertyTokenizer from "./boolean"
import expressionValueTokenizer from "./expression"
import keyValueTokenizer from "./key"
import literalValueTokenizer from "./literal"
import SourceStream from "./sourceStream"
import tagCloseTokenizer from "./tagClose"

import { tagOpenBeginTokenizer, tagOpenEndTokenizer } from "./tagOpen"
import tagPropertyTokenizer from "./tagProperty"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"

enum TokenizerMode {
    Contents,
    Tag
}

export const tokenizer = (sourceStream: SourceStream): Token[] => {
    let mode: TokenizerMode = TokenizerMode.Contents
    let currentDescription: TokenDescription | undefined
    let returnValue: Token[] = []

    while(!sourceStream.isEndOfSource) {
        if (mode === TokenizerMode.Contents) {
            const checkSubTokens = checkSubTokenizers<TokenWhitespace | TokenTagClose | TokenTagOpenBegin>({
                sourceStream,
                subTokenizers: [whiteSpaceTokenizer, tagCloseTokenizer, tagOpenBeginTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenBegin') {
                        mode = TokenizerMode.Tag
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                source: sourceStream.lookAhead(1),
                message: 'Unexpected token'
            }]
        }
        else {
            const checkSubTokens = checkSubTokenizers<TokenTagOpenEnd | TokenWhitespace | TokenProperty | TokenBooleanProperty>({
                sourceStream,
                subTokenizers: [tagOpenEndTokenizer, whiteSpaceTokenizer, tagPropertyTokenizer, booleanPropertyTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenEnd') {
                        mode = TokenizerMode.Contents
                    }
                    if (token.type === 'Property') {
                        //
                        // Consume any whitespace first...
                        //
                        const checkWhitespace = whiteSpaceTokenizer(sourceStream)
                        if (checkWhitespace) {
                            if (isTokenError(checkWhitespace)) {
                                return {
                                    success: false,
                                    error: checkWhitespace
                                }
                            }
                            returnValue.push(checkWhitespace)
                        }
                        //
                        // ... and then confirm that a value follows the key
                        //
                        const checkValue = checkSubTokenizers<TokenLiteralValue | TokenKeyValue | TokenExpressionValue>({
                            sourceStream,
                            subTokenizers: [literalValueTokenizer, keyValueTokenizer, expressionValueTokenizer],
                            callback: (token) => {
                                returnValue.push(token)
                            }
                        })
                        if (checkValue) {
                            if (checkValue.success === false) {
                                return checkValue
                            }
                        }
                        else {
                            return {
                                success: false,
                                error: {
                                    type: 'Error',
                                    source: sourceStream.lookAhead(1),
                                    startIdx: sourceStream.position,
                                    endIdx: sourceStream.position,
                                    message: 'Unexpected token'
                                }
                            }
                        }
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                source: sourceStream.lookAhead(1),
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                message: 'Unexpected token'
            }]
        }
    }
    return returnValue
}

export default tokenizer
