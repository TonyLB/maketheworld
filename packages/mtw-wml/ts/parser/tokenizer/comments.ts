import { Tokenizer, TokenComment } from "./baseClasses"

export const commentTokenizer: Tokenizer<TokenComment> = (sourceStream) => {
    if (sourceStream.lookAhead('//')) {
        const nextInstance = sourceStream.nextInstance(['\n'])
        const startIdx = sourceStream.position
        if (nextInstance > 1) {
            sourceStream.consume(nextInstance)
        }
        else {
            sourceStream.consume(sourceStream.source.length - sourceStream.position)
        }
        const endIdx = sourceStream.position - 1
        return {
            type: 'Comment',
            startIdx,
            endIdx
        }
    }
    if (sourceStream.lookAhead('/*')) {
        const nextInstance = sourceStream.nextInstance(['*/'])
        const startIdx = sourceStream.position
        if (nextInstance > 1) {
            sourceStream.consume(nextInstance + 2)
        }
        else {
            sourceStream.consume(sourceStream.source.length - sourceStream.position)
        }
        const endIdx = sourceStream.position - 1
        if (nextInstance === -1) {
            return {
                type: 'Error',
                message: 'Unbounded comment',
                startIdx,
                endIdx
            }
        }
        else {
            return {
                type: 'Comment',
                startIdx,
                endIdx
            }
        }
    }
    return undefined
}

export default commentTokenizer
