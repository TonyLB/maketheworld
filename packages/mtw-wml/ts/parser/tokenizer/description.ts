import { Tokenizer, TokenDescription } from "./baseClasses"

export const descriptionTokenizer: Tokenizer<TokenDescription> = (sourceStream) => {
    const startIdx = sourceStream.position
    let value = ''
    while (![' ', '\n', '\t', '\r', '<'].includes(sourceStream.lookAhead(1))) {
        if (['/*', '//'].includes(sourceStream.lookAhead(2))) {
            break
        }
        if (sourceStream.isEndOfSource) {
            break
        }
        if (sourceStream.lookAhead(1) === '\\') {
            sourceStream.consume(1)
        }
        value = `${value}${sourceStream.consume(1)}`
    }
    const endIdx = sourceStream.position
    if (startIdx === endIdx) {
        return undefined
    }
    else {
        return {
            type: 'Description',
            startIdx,
            endIdx,
            value
        }
    }
}

export default descriptionTokenizer
