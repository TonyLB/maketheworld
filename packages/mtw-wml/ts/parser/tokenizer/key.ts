import { Tokenizer, TokenKeyValue } from "./baseClasses"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"

export const keyValueTokenizer: Tokenizer<TokenKeyValue> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar === '(') {
        const startIdx = sourceStream.position
        sourceStream.consume(1)
        const checkWhitespaceOne = checkSubTokenizers({
            sourceStream,
            subTokenizers: [whiteSpaceTokenizer],
            callback: (token) => {}
        })
        if (checkWhitespaceOne && (checkWhitespaceOne.success === false)) {
            return checkWhitespaceOne.error
        }
        const valueStartIdx = sourceStream.position
        if (sourceStream.lookAhead(1).match(/[A-Za-z\_]/)) {
            sourceStream.consume(1)
            while(sourceStream.lookAhead(1).match(/[A-Za-z0-9\_]/)) {
                sourceStream.consume(1)
            }    
        }
        const value = sourceStream.source.slice(valueStartIdx, sourceStream.position)
        const checkWhitespaceTwo = checkSubTokenizers({
            sourceStream,
            subTokenizers: [whiteSpaceTokenizer],
            callback: (token) => {}
        })
        if (checkWhitespaceTwo && (checkWhitespaceTwo.success === false)) {
            return checkWhitespaceTwo.error
        }
        if (!sourceStream.lookAhead(')')) {
            return {
                type: 'Error',
                source: sourceStream.lookAhead(1),
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                message: 'Unexpected token'
            }
        }
        sourceStream.consume(1)
        if (sourceStream.position !== startIdx) {
            return {
                type: 'KeyValue',
                value,
                startIdx,
                endIdx: sourceStream.position - 1
            }
        }
    }
    return undefined
}

export default keyValueTokenizer
