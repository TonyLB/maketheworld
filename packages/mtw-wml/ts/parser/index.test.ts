import parse from '.'
import { ParseException } from './baseClasses'
import tokenizer from './tokenizer'
import SourceStream from './tokenizer/sourceStream'

describe('wml parser', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should return empty list from no tokens', () => {
        expect(parse([])).toEqual([])
    })
    it('should parse a single tag', () => {
        const testTokens = tokenizer(new SourceStream('<Asset key=(Test)></Asset>'))
        expect(parse(testTokens)).toMatchSnapshot()
    })
    it('should ignore whitespace outside tags', () => {
        const testTokens = tokenizer(new SourceStream('    <Asset key=(Test)></Asset>\n    '))
        expect(parse(testTokens)).toMatchSnapshot()
    })
    it('should parse one level of nesting', () => {
        const testTokens = tokenizer(new SourceStream('<Asset key=(Test)><Room key=(ABC) /></Asset>'))
        expect(parse(testTokens)).toMatchSnapshot()
    })
    it('should parse elements correctly', () => {
        const testTokens = tokenizer(new SourceStream(`
            <Asset key=(Test) fileName="test">
                <Import from=(BASE)>
                    <Use key=(basePower) type="Variable" as=(power) />
                    <Use key=(overview) type="Room" />
                </Import>
                <Room key=(ABC)>
                    <Name>Vortex</Name>
                    <Description>
                        Vortex
                        <Link to=(toggleOpen)>(toggle)</Link>
                    </Description>
                    <Exit from=(DEF)>vortex</Exit>
                </Room>
                <Condition if={open}>
                    <Depend on=(open) />
                    <Room key=(ABC)>
                        <Exit to=(DEF)>welcome</Exit>
                    </Room>
                </Condition>
                <Room key=(DEF)>
                    <Name>Welcome</Name>
                </Room>
                <Variable key=(open) default={false} />
                <Action key=(toggleOpen) src={open = !open} />
                <Computed key=(closed) src={!open}>
                    <Depend on=(open) />
                </Computed>
            </Asset>
        `))
        expect(parse(testTokens)).toMatchSnapshot()
    })
    it('should error on illegal contents', () => {
        const testTokens = tokenizer(new SourceStream('<Asset key=(Test)><Link to=(ABC)>test</Link></Asset>'))
        expect(() => {
            parse(testTokens)
        }).toThrow(ParseException)
    })
    it('should parse a character tag correctly', () => {
        const testTokens = tokenizer(new SourceStream(`
            <Character key=(Tess) fileName="Tess" player="testy" zone="Library">
                <Image key=(icon) fileURL="testIcon.png" />
                <Name>Tess</Name>
                <Pronouns
                    subject="she"
                    object="her"
                    possessive="her"
                    adjective="hers"
                    reflexive="herself"
                />
            </Character>
        `))
        expect(parse(testTokens)).toMatchSnapshot()
    })

})