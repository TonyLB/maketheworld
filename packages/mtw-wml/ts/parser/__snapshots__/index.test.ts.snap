// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wml parser should ignore whitespace outside tags 1`] = `
Array [
  Object {
    "contents": Array [],
    "endTagToken": 5,
    "key": "Test",
    "startTagToken": 0,
    "tag": "Asset",
  },
]
`;

exports[`wml parser should parse a character tag correctly 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "endTagToken": 22,
        "fileURL": "testIcon.png",
        "key": "icon",
        "startTagToken": 14,
        "tag": "Image",
      },
      Object {
        "contents": Array [
          Object {
            "endTagToken": 25,
            "startTagToken": 25,
            "tag": "String",
            "value": "Tess",
          },
        ],
        "endTagToken": 26,
        "startTagToken": 23,
        "tag": "Name",
        "value": "Tess",
      },
      Object {
        "adjective": "hers",
        "endTagToken": 44,
        "object": "her",
        "possessive": "her",
        "reflexive": "herself",
        "startTagToken": 27,
        "subject": "she",
        "tag": "Pronouns",
      },
    ],
    "endTagToken": 45,
    "fileName": "Tess",
    "key": "Tess",
    "player": "testy",
    "startTagToken": 0,
    "tag": "Character",
    "zone": "Library",
  },
]
`;

exports[`wml parser should parse a map tag correctly 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "endTagToken": 16,
                "startTagToken": 16,
                "tag": "String",
                "value": "Test Map",
              },
            ],
            "endTagToken": 18,
            "startTagToken": 14,
            "tag": "Name",
            "value": "Test Map",
          },
          Object {
            "endTagToken": 27,
            "fileURL": "https://test.com/imageTest.png",
            "key": "ImageTest",
            "startTagToken": 19,
            "tag": "Image",
          },
          Object {
            "contents": Array [],
            "endTagToken": 39,
            "key": "ABC",
            "startTagToken": 28,
            "tag": "Room",
            "x": 200,
            "y": 150,
          },
        ],
        "endTagToken": 40,
        "key": "TestMap",
        "startTagToken": 9,
        "tag": "Map",
      },
    ],
    "endTagToken": 41,
    "fileName": "test",
    "instance": true,
    "key": "Test",
    "startTagToken": 0,
    "tag": "Story",
  },
]
`;

exports[`wml parser should parse a single tag 1`] = `
Array [
  Object {
    "contents": Array [],
    "endTagToken": 5,
    "key": "Test",
    "startTagToken": 0,
    "tag": "Asset",
  },
]
`;

exports[`wml parser should parse a story tag correctly 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "endTagToken": 16,
                "startTagToken": 16,
                "tag": "String",
                "value": "Vortex",
              },
            ],
            "endTagToken": 17,
            "startTagToken": 14,
            "tag": "Name",
            "value": "Vortex",
          },
          Object {
            "contents": Array [
              Object {
                "endTagToken": 20,
                "startTagToken": 20,
                "tag": "String",
                "value": "Vortex",
              },
            ],
            "endTagToken": 21,
            "spaceAfter": false,
            "spaceBefore": false,
            "startTagToken": 18,
            "tag": "Description",
          },
        ],
        "endTagToken": 22,
        "key": "ABC",
        "startTagToken": 9,
        "tag": "Room",
        "x": undefined,
        "y": undefined,
      },
    ],
    "endTagToken": 23,
    "fileName": "test",
    "instance": true,
    "key": "Test",
    "startTagToken": 0,
    "tag": "Story",
  },
]
`;

exports[`wml parser should parse elements correctly 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "as": "power",
            "endTagToken": 24,
            "key": "basePower",
            "startTagToken": 13,
            "tag": "Use",
            "type": "Variable",
          },
          Object {
            "endTagToken": 33,
            "key": "overview",
            "startTagToken": 25,
            "tag": "Use",
            "type": "Room",
          },
        ],
        "endTagToken": 34,
        "from": "BASE",
        "startTagToken": 8,
        "tag": "Import",
      },
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "endTagToken": 42,
                "startTagToken": 42,
                "tag": "String",
                "value": "Vortex",
              },
            ],
            "endTagToken": 43,
            "startTagToken": 40,
            "tag": "Name",
            "value": "Vortex",
          },
          Object {
            "contents": Array [
              Object {
                "endTagToken": 46,
                "startTagToken": 46,
                "tag": "String",
                "value": "Vortex",
              },
              Object {
                "endTagToken": 47,
                "startTagToken": 47,
                "tag": "Whitespace",
              },
              Object {
                "contents": Array [
                  Object {
                    "endTagToken": 53,
                    "startTagToken": 53,
                    "tag": "String",
                    "value": "(toggle)",
                  },
                ],
                "endTagToken": 54,
                "startTagToken": 48,
                "tag": "Link",
                "to": "toggleOpen",
              },
            ],
            "endTagToken": 55,
            "spaceAfter": false,
            "spaceBefore": false,
            "startTagToken": 44,
            "tag": "Description",
          },
          Object {
            "contents": Array [
              Object {
                "endTagToken": 61,
                "startTagToken": 61,
                "tag": "String",
                "value": "vortex",
              },
            ],
            "endTagToken": 62,
            "from": "DEF",
            "startTagToken": 56,
            "tag": "Exit",
          },
        ],
        "endTagToken": 63,
        "key": "ABC",
        "startTagToken": 35,
        "tag": "Room",
        "x": undefined,
        "y": undefined,
      },
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "contents": Array [
                  Object {
                    "endTagToken": 85,
                    "startTagToken": 85,
                    "tag": "String",
                    "value": "welcome",
                  },
                ],
                "endTagToken": 86,
                "startTagToken": 80,
                "tag": "Exit",
                "to": "DEF",
              },
            ],
            "endTagToken": 87,
            "key": "ABC",
            "startTagToken": 75,
            "tag": "Room",
            "x": undefined,
            "y": undefined,
          },
        ],
        "dependencies": Array [
          Object {
            "endTagToken": 74,
            "on": "open",
            "startTagToken": 69,
            "tag": "Depend",
          },
        ],
        "endTagToken": 88,
        "if": "open",
        "startTagToken": 64,
        "tag": "Condition",
      },
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "endTagToken": 96,
                "startTagToken": 96,
                "tag": "String",
                "value": "Welcome",
              },
            ],
            "endTagToken": 97,
            "startTagToken": 94,
            "tag": "Name",
            "value": "Welcome",
          },
        ],
        "endTagToken": 98,
        "key": "DEF",
        "startTagToken": 89,
        "tag": "Room",
        "x": undefined,
        "y": undefined,
      },
      Object {
        "default": "false",
        "endTagToken": 107,
        "key": "open",
        "startTagToken": 99,
        "tag": "Variable",
      },
      Object {
        "endTagToken": 116,
        "key": "toggleOpen",
        "src": "open = !open",
        "startTagToken": 108,
        "tag": "Action",
      },
      Object {
        "dependencies": Array [
          Object {
            "endTagToken": 130,
            "on": "open",
            "startTagToken": 125,
            "tag": "Depend",
          },
        ],
        "endTagToken": 131,
        "key": "closed",
        "src": "!open",
        "startTagToken": 117,
        "tag": "Computed",
      },
    ],
    "endTagToken": 132,
    "fileName": "test",
    "key": "Test",
    "startTagToken": 0,
    "tag": "Asset",
  },
]
`;

exports[`wml parser should parse one level of nesting 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [],
        "endTagToken": 10,
        "key": "ABC",
        "startTagToken": 5,
        "tag": "Room",
        "x": undefined,
        "y": undefined,
      },
    ],
    "endTagToken": 11,
    "key": "Test",
    "startTagToken": 0,
    "tag": "Asset",
  },
]
`;
